### PR TITLE
Pin ``sphinx-autodoc-typehint``

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

The 1.23.4 release of ``sphinx-autodoc-typehint`` broke the docs build of the CI.

### Details and comments

The build errors with
```
reading sources... [  3%] apidocs/algorithms .. stubs/qiskit.algorithms.SciPyImaginaryEvolver
:26: (ERROR/3) no path specified
:26: (ERROR/3) no path specified

Traceback (most recent call last):
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.9/site-packages/sphinx/cmd/build.py", line 281, in build_main
    app.build(args.force_all, args.filenames)
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.9/site-packages/sphinx/application.py", line 347, in build
    self.builder.build_update()
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 310, in build_update
    self.build(to_build,
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 326, in build
    updated_docnames = set(self.read())
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 431, in read
    self._read_parallel(docnames, nproc=self.app.parallel)
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 487, in _read_parallel
    tasks.join()
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.9/site-packages/sphinx/util/parallel.py", line 105, in join
    if not self._join_one():
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.9/site-packages/sphinx/util/parallel.py", line 126, in _join_one
    raise SphinxParallelError(*result)
sphinx.errors.SphinxParallelError: sphinx.errors.ExtensionError: Handler <function process_docstring at 0x7f914bf254c0> for event 'autodoc-process-docstring' threw an exception (exception: )

Sphinx parallel build error:
sphinx.errors.ExtensionError: Handler <function process_docstring at 0x7f914bf254c0> for event 'autodoc-process-docstring' threw an exception (exception: )
docs: exit 2 (43.29 seconds) /home/vsts/work/1/s> sphinx-build -W -j auto -T --keep-going -b html docs/ docs/_build/html pid=5682
.pkg: _exit> python /opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/pyproject_api/_backend.py True setuptools.build_meta
  docs: FAIL code 2 (244.43=setup[201.14]+cmd[43.29] seconds)
  evaluation failed :( (245.89 seconds)
##[error]Bash exited with code '2'.
```
see e.g. https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=42940&view=logs&j=abfb3e93-d1ff-5b1b-cb4c-74d6dcf536f9&t=60edc4b3-4259-59c8-ae13-dc31febe2d3d

Let's see if pinning to the latest version before that, 1.23.3, works otherwise we can pin 1.23.1 which worked before.

### CODE REVIEW

**Comment by jakelishman** [comment]
You want 1.21.1 as the pin, most likely. I think the maintainer made a mistake and jumped straight from 1.21.3 to 1.23.4.

1.21.1 is known good for us.

**Comment by HuangJunye** [comment]
> You want 1.21.1 as the pin, most likely. I think the maintainer made a mistake and jumped straight from 1.21.3 to 1.23.4.
> 
> 1.21.1 is known good for us.

OMG. I am not sure I would ever remove the pin for it. The maintainer is so careless...

**Comment by Cryoris** [comment]
Done in 401906d 👍🏻

File: requirements-dev.txt [hunk 1]
```diff
@@ -18,7 +18,7 @@ seaborn>=0.9.0
 reno>=3.4.0
 Sphinx>=3.0.0
 qiskit-sphinx-theme>=1.6
-sphinx-autodoc-typehints>=1.18
+sphinx-autodoc-typehints==1.23.1  # revert to >=1.18 once the incompatibilities in >1.23.1 are fixed
```
Comment (HuangJunye) [inline requirements-dev.txt:21]:
```suggestion
sphinx-autodoc-typehints==1.21.1  # revert to >=1.18 once the incompatibilities in >1.23.1 are fixed
```
It should be `1.21.1` not `1.23.1`. See https://github.com/Qiskit/qiskit-terra/pull/9383#issuecomment-1386778745

Comment (Cryoris) [reply requirements-dev.txt]: done in 8a941c9

**Comment by jakelishman** [comment]
Depending on the release process, messing up the version is a pretty easy mistake when you're just one guy maintaining a small package that's seen a sudden big uptick in activity.

For repeatable builds, though, I'm moving more in favour of us having a completely frozen (dependencies and transitives) "known good" Python environment to build documentation in, on CI at least.  It could add friction for users if our _only_ development requirements file pins absolutely everything, and it's a bit harder to maintain because raw `pip` doesn't really give us the tools to separate hard from transitive dependencies in that style, but I'll think about it.

**Review by jakelishman — APPROVED** [review]

**Comment by mtreinish** [comment]
The easiest way to do a frozen environment like that is to use a constraints file to pin the full version. If you have a fixed version for the full pip freeze of a working run in a constraints file pip will reconstruct the same exact environment for us. In a previous project that's what we did for all dependencies (globally across all projects) and then had an automated process to try bumping things when there were new releases to test for regressions before starting ci. It wasn't a perfect solution but it did work to isolate some of the failures from broken upstream dependency releases.

**Pull request merged**